cmd/k8s-operator/hostscm.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+// tailscale-operator provides a way to expose services running in a Kubernetes
+// cluster to your Tailnet.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"tailscale.com/client/tailscale/apitype"
+)
+
+const dnsConfigKey = "dns.json"
+
+// hostsCMConfig contains the config needed to update ts.net config for a
+// particular egress proxy
+type hostsCMConfig struct {
+	// IP of the Tailscale service that we are setting up egress for
+	targetIP string
+	// serviceLabels identify the proxy service
+	serviceLabels map[string]string
+}
+
+// hostsCMProvisioner knows how to update ts.net nameserver config as egress
+// proxy Services get added/deleted
+type hostsCMProvisioner struct {
+	// Client reads from cache first, then kube-apiserver, writes to
+	// kube-apiserver
+	client.Client
+	// namespace in which tailscale resources get provisioned
+	tsNamespace string
+	// localClient knows how to talk to tailscaled local API
+	localAPIClient localClient
+}
+
+type localClient interface {
+	WhoIs(ctx context.Context, remoteAddr string) (*apitype.WhoIsResponse, error)
+}
+
+// Provision ensures that ts.net nameserver config has been updated with a
+// mapping of Tailscale service to its egress proxy kube Service IP. If
+// successful, returns the Tailscale service FQDN
+func (h *hostsCMProvisioner) Provision(ctx context.Context, logger *zap.SugaredLogger, hcc *hostsCMConfig) (string, error) {
+	logger.Info("IN CM PROV")
```

Comment:
Do you want to merge this eventually? If yes, perhaps it's worth logging the contents of `hcc` as well?

Reply:
Thanks, this needs to be removed